### PR TITLE
Refactoring after refactoring 

### PR DESCRIPTION
**Pull request by MaksimZhukov**

### CODE REVIEW

File: azure-pipelines/test_job.yml [hunk 1]
```diff
@@ -3,7 +3,13 @@ jobs:
   pool: 
     name: Azure Pipelines
     vmImage: $(VmImage)
-  
+  strategy:
+    matrix:
+      Cleanup-Python-Toolcache-Dir:
+        IsCleanupMode: true
```
Comment (maxim-lobanov) [inline azure-pipelines/test_job.yml]: NeedCleanToolcacheDir

File: azure-pipelines/test_job.yml [hunk 2]
```diff
@@ -3,7 +3,13 @@ jobs:
   pool: 
     name: Azure Pipelines
     vmImage: $(VmImage)
-  
+  strategy:
+    matrix:
+      Cleanup-Python-Toolcache-Dir:
```
Comment (maxim-lobanov) [inline azure-pipelines/test_job.yml]: validate-python-installation

File: azure-pipelines/test_job.yml [hunk 3]
```diff
@@ -3,7 +3,13 @@ jobs:
...
+        IsCleanupMode: true
+      Invoke-Pester-Tests:
```
Comment (maxim-lobanov) [inline azure-pipelines/test_job.yml]: validate-python-re-installation

File: azure-pipelines/test_job.yml [hunk 4]
```diff
@@ -65,4 +99,4 @@ jobs:
       testResultsFormat: NUnit
       searchFolder: 'tests'
       failTaskOnFailedTests: true
-    condition: always()
+    condition: ne(variables['IsCleanupMode'], true)
```
Comment (maxim-lobanov) [inline azure-pipelines/test_job.yml]: Why we don't publish tests if `IsCleanupMode` is true?
How we will determine that tests failed?

File: azure-pipelines/test_job.yml [hunk 5]
```diff
@@ -26,12 +32,39 @@ jobs:
...
+          Get-WmiObject Win32_Product -Filter $PythonFilter | Foreach-Object { 
+            Write-Host "Uninstalling $($_.Name) ..."
+            $_.Uninstall() | Out-Null 
+          }
+        }
+        $PythonToolcachePath = Join-Path -Path $env:AGENT_TOOLSDIRECTORY -ChildPath "Python"
+        Write-Host "Remove Python toolcache folder"
+        Remove-Item -Path $PythonToolcachePath -Recurse -Force
+      workingDirectory: '$(Build.BinariesDirectory)'
+    condition: eq(variables['IsCleanupMode'], true)
```
Comment (maxim-lobanov) [inline azure-pipelines/test_job.yml]: I think clean up should be the first one in the yml.
Also, can we avoid much logic here? Why we need the separate logic for Windows? Will just tool-cache dir clean up handle the cases that we have discussed earlier?

File: azure-pipelines/test-steps.yml [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  inputs:
+    versionSpec: '$(VERSION)'
+    architecture: '$(Architecture)'
+
+- task: PowerShell@2
+  displayName: Check shared libraries
+  inputs:
+    TargetType: inline
+    script: |
+      $pythonLocation = "$Env:USEPYTHONVERSION_PYTHONLOCATION/bin/python"
```
Comment (vmapetr) [inline azure-pipelines/test-steps.yml]: Why not simple `which python`?

**Pull request merged**